rfcs/01391_async_sdk.md

```diff
@@ -0,0 +1,189 @@
+    F: FnOnce(GrpcStream<T>) -> R,
+    R: Future<Output = ()> + 'static,
+```
+
+The `TryFrom` implementation is automatically generated by the service generator.
+
+`MultiWriter`  can be used to send 0 or more messages:
+
+```rust
+impl<T: Message> MultiWriter {
+    // Note: self by reference
+    pub fn send(&self, msg: &T) -> Result<(), OakError> { /* ... */ }
+}
+
+fn use_multi(w: MultiWriter<HelloResponse>) {
+    w.send(HelloResponse::default()).unwrap();
+    w.send(HelloResponse::default()).unwrap();
+    // Automatically close channel on `Drop`
+}
+```
+
+The `OneshotWriter` is similar, but only allows a single response to be written:
+
+```rust
+impl<T: Message> OneshotWriter {
+    // Note: self by value
+    pub fn send(self, msg: &T) -> Result<(), OakError> { /* ... */ }
+}
+
+fn use_oneshot(w: OneshotWriter<HelloResponse>) {
+    w.send(HelloResponse::default()).unwrap();
+    // `w` can no longer be used.
+    // channel is automatically closed.
+}
+```
+
+## Executor and primitive internals
+When a `ChannelRead` is polled and no data is readily available, it must request the executor to wake it once data does become available. We take a similar approach to other executors like [`tokio`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/context.rs#L7) here, and make the executor available to `ChannelRead` via thread-local storage (TLS). Note that for Webassembly this is equivalent to a global variable, since threads are not standardized or widely supported. While the initial version of the executor will be single-threaded, the API is designed to allow for multithreaded execution.
+
+The executor maintains a set of [`Waker`](https://doc.rust-lang.org/std/task/struct.Waker.html) objects and the channels associated with them, which the `ChannelRead` futures add themselves to when polled. The executor runs the following loop:
+```rust
+loop {
+    // futures_pool: Set of the main event loop future and all `spawn`ed futures.
+    // channels: Set of all channels that have wakers associated with them. Populated by `ChannelRead`.
+    // wakers: Set of all wakers. Populated by `ChannelRead`.
+
+    // Returns when no future can make progress
+    futures_pool.run_until_stalled();
+
+    if futures_pool.is_empty() {
+        break;
+    }
+
+    let channel_read_statuses = wait_on_channels(&channels);
+
+    // For each ready channel, awake one of the futures that was waiting on it.
+    wake_if_ready(channel_read_statuses, wakers);
+}
+``` 
+
+Implementing an executor for scratch is hard, unsafe and error-prone, even for a [very simple basic](https://github.com/spacejam/extreme#bugs-encountered-over-time). Instead, we can use an existing and tested [futures pool](https://docs.rs/futures/0.3.5/futures/executor/struct.LocalPool.html) that is part of the rust-lang supported `futures` crate, and implement our executor on top of that. This greatly reduces engineering effort, unsafe code and testing surface.
```

Comment:
I am still unclear as to how polling a future actually results in a `wait_on_channels` call at the bottom of the stack, do you have some more details to provide?

Reply:
I'll try and explain it here in a comment, please let me know if you think it would be useful to put in the RFC:

When a `ChannelRead` is polled and finds the channel not ready, it provides the executor with a `(Handle, Waker)`. The executor maintains a global `HashMap<Handle, Vec<Waker>>` that this tuple is then added to.

If at some point none of the futures managed by the executor can make any more progress, the executor issues a `wait_on_channels` call for all handles stores in the global map. 
For each handle found ready, the executor removes one `Waker` from the map and uses it to wake up the future. Once woken up, the future issues a `channel_read` and retrieves the data.

I hope that makes sense 😄

Reply:
> When a `ChannelRead` is polled and finds the channel not ready, it provides the executor with a `(Handle, Waker)`. The executor maintains a global `HashMap<Handle, Vec<Waker>>` that this tuple is then added to.

I guess this is the part that is not fully clear to me, how does `ChannelRead` pass this information to the executor? Global mutable state? Or some other magic mechanism?

Reply:
Global mutable state unfortunately. This is actually the approach all other executors I've seen take, and I guess for an executor it's not too bad.

Reply:
Ok makes sense I guess, could you mention it in the doc?